### PR TITLE
[enhancement] ConvBinWinogradRxSf2x3 / f3x2 solvers merge

### PR DESCRIPTION
**Pull request by johnny-keker**
This PR merges almost similar `ConvBinWingradRxSf2x3` and `ConvBinWingradRxSf3x2` solvers into one `ConvBinWinoRxS` solver, which reduces the amount of duplicated code. In order to provide similar functionality, new solver uses templates:

```
    ConvBinWinoRxS<2, 3>;
    ConvBinWinoRxS<3, 2>;
```

@atamazov please review

### CODE REVIEW

**Comment by junliume** [comment]
@Slimakanzer We are having some issues with Perf DB Deserialize Test. I think it is because of the merge of two solvers.
@JehandadKhan maybe the DB should be updated?

**Review by atamazov — CHANGES_REQUESTED** [review]
If a solver is renamed, then
- correctness: the affected solver must be registered with different Solver Id.
- performance: find-db must be updated at least. If solver is tunable, then perf-db must be updated too.

File: src/solver.cpp [hunk 1]
```diff
@@ -303,7 +303,7 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
     RegisterWithSolver(
         registry, ++id, ConvWinograd3x3MultipassWrW<3, 4>{}, miopenConvolutionAlgoWinograd);
     ++id; // Id for ConvSCGemmFGemm.
-    RegisterWithSolver(registry, ++id, ConvBinWinogradRxSf3x2{}, miopenConvolutionAlgoWinograd);
```
Comment (atamazov) [inline src/solver.cpp:306]: here

File: src/solver.cpp [hunk 2]
```diff
@@ -333,7 +333,7 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
     ++id; // removed solver ConvHipImplicitGemmV4R4GenFwdXdlops
     ++id; // removed solver ConvHipImplicitGemmV4R4GenWrWXdlops
 
-    RegisterWithSolver(registry, ++id, ConvBinWinogradRxSf2x3{}, miopenConvolutionAlgoWinograd);
```
Comment (atamazov) [inline src/solver.cpp:336]: and here

**Comment by atamazov** [comment]
@johnny-keker BTW you have to update doc/DebuggingAndLogging.md

Or you can implement custom GetSolverDbId() functions and keep solver names intact. I think this is the best solution.

**Comment by johnny-keker** [comment]
> @johnny-keker BTW you have to update doc/DebuggingAndLogging.md
> ### green_circle Or you can implement custom GetSolverDbId() functions and keep solver names intact. I think this is the best solution.

Thanks for the suggestion! Implemented in [74c72f6](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1533/commits/74c72f64ecd7939b7272d00bab0fac7f46d7f016)

File: src/solver/conv_winoRxS.cpp [hunk 3]
```diff
@@ -467,7 +476,7 @@ static bool IsApplicableBase(const ConvolutionContext& params)
         return false;
 
     // clang-format off
-    if (! ( (params.kernel_stride_w == 1 || params.kernel_stride_w == 2)
+    if (! (params.kernel_stride_w == 1
```
Comment (Slimakanzer) [inline src/solver/conv_winoRxS.cpp:479]: Please check these changes. 
It means that the case is applicable only with `kernel_stride=1`, even if it was _F2x3_ solver.

File: src/solver/conv_winoRxS.cpp [hunk 4]
```diff
@@ -513,17 +522,65 @@ static bool IsApplicableBase(const ConvolutionContext& params)
...
+        env_name = MIOPEN_DEBUG_AMD_WINOGRAD_RXS_F2X3_PERF_VALS::value();
+    }
+    else if(IS3x2)
+    {
+        p_asciz  = miopen::GetStringEnv(MIOPEN_DEBUG_AMD_WINOGRAD_RXS_F3X2_PERF_VALS{});
+        env_name = MIOPEN_DEBUG_AMD_WINOGRAD_RXS_F3X2_PERF_VALS::value();
+    }
+
+    if(p_asciz == nullptr)
+        return boost::none;
```
Comment (atamazov) [inline src/solver/conv_winoRxS.cpp:561]: [Recommendation] This should construct an instance of `boost::optional<PerformanceConfigConvBinWinogradRxS>` that is empty:
```suggestion
        return {};
```

File: src/solver/conv_winoRxS.cpp [hunk 5]
```diff
@@ -540,30 +597,15 @@ ConvBinWinogradRxSf2x3::GetSolution(const ConvolutionContext& params,
...
-    const PerformanceConfigConvBinWinogradRxSf2x3* pcfg = &config;
-    PerformanceConfigConvBinWinogradRxSf2x3 fromEnv;
+    const PerformanceConfigConvBinWinogradRxS* pcfg = &config;
 
     if(!disableConfigOverrideFromEnv)
     {
-        std::string s;
-        const auto p_asciz = miopen::GetStringEnv(MIOPEN_DEBUG_AMD_WINOGRAD_RXS_F2X3_PERF_VALS{});
-        if(p_asciz != nullptr)
+        boost::optional<PerformanceConfigConvBinWinogradRxS> fromEnv =
```
Comment (atamazov) [inline src/solver/conv_winoRxS.cpp:604]:
```suggestion
        const auto fromEnv =
```

File: src/solver/conv_winoRxS.cpp [hunk 6]
```diff
@@ -600,21 +642,34 @@ ConvBinWinogradRxSf2x3::GetSolution(const ConvolutionContext& params,
...
     }
-    else // if(params.kernel_dilation_h == 2)
+    else if(IS3x2)
     {
-        kernel_postfix += "_dilation2";
-    }
-
-    if(params.group_counts != 1 || params.direction.IsBackwardWrW())
+        kernel_postfix = params.IsFp32() ? "_f3x2_fp32" : "_f3x2_fp16_dot2_edc";
+        // if(params.kernel_stride_w == 1)
```
Comment (atamazov) [inline src/solver/conv_winoRxS.cpp:669]: Do we need this comment?

**Review by atamazov — CHANGES_REQUESTED** [review]
Almost good, but no go yet.

Please also resolve https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1533#discussion_r864978898, that comment is blocking.

**Comment by atamazov** [comment]
@johnny-keker Can you please also describe how the refactored solvers were tested?

**Review by junliume — APPROVED** [review]

**Review by junliume — CHANGES_REQUESTED** [review]
Second thoughts, maybe not quite there yet. I am checking CI pipelines again.

**Comment by junliume** [comment]
@Slimakanzer  and @johnny-keker CI have passed. However, there are some merge conflicts. Could you please resolve them and try to merge this PR?

**Comment by johnny-keker** [comment]
@junliume I resolved the conflicts, and CI seems to be passed. Should we consider merging?

**Review by junliume — APPROVED** [review]

**Pull request merged**